gpt4all-chat/database.cpp

```diff
@@ -990,47 +1265,107 @@ bool Database::addFolder(const QString &collection, const QString &path, bool fr
         qWarning() << "ERROR: Cannot add folder to db with path" << path << q.lastError();
-        return false;
+        return;
+    }
+
+    Q_ASSERT(folder_id != -1);
+
+    if (!addCollection(q, collection.collection, folder_id,
+                       QDateTime() /*last_update*/,
+                       m_embLLM->model() /*embedding_model*/,
+                       true /*force_indexing*/)) {
+        qWarning() << "ERROR: Cannot add folder to collection" << collection.collection << path << q.lastError();
+        return;
+    }
+
+    addGuiCollectionItem(collection);
+}
+
+void Database::forceIndexing(const QString &collection)
+{
+    QSqlQuery q;
+    QList<QPair<int, QString>> folders;
+    if (!selectFoldersFromCollection(q, collection, &folders)) {
+        qWarning() << "ERROR: Cannot select folders from collections" << collection << q.lastError();
+        return;
+    }
+
+    if (!updateCollectionForceIndexing(q, collection)) {
+        qWarning() << "ERROR: Cannot update collection" << collection << q.lastError();
+        return;
+    }
+
+    for (const auto& folder : folders) {
+        CollectionItem item = guiCollectionItem(folder.first);
+        item.forceIndexing = false;
+        updateGuiForCollectionItem(item);
+        addFolder(collection, folder.second);
+    }
+}
+
+bool containsFolderId(const QList<QPair<int, QString>> &folders, int folder_id) {
+    for (const auto& folder : folders)
+        if (folder.first == folder_id)
+            return true;
+    return false;
+}
+
+void Database::addFolder(const QString &collection, const QString &path)
+{
+    QFileInfo info(path);
+    if (!info.exists() || !info.isReadable()) {
+        qWarning() << "ERROR: Cannot add folder that doesn't exist or not readable" << path;
+        return;
+    }
+
+    QSqlQuery q;
+    int folder_id = -1;
+
+    // See if the folder exists in the db
+    if (!selectFolder(q, path, &folder_id)) {
+        qWarning() << "ERROR: Cannot select folder from path" << path << q.lastError();
+        return;
+    }
+
+    // Add the folder
+    if (folder_id == -1 && !addFolderToDB(q, path, &folder_id)) {
+        qWarning() << "ERROR: Cannot add folder to db with path" << path << q.lastError();
+        return;
     }
 
     Q_ASSERT(folder_id != -1);
 
     // See if the folder has already been added to the collection
-    QList<int> folders;
+    QList<QPair<int, QString>> folders;
     if (!selectFoldersFromCollection(q, collection, &folders)) {
         qWarning() << "ERROR: Cannot select folders from collections" << collection << q.lastError();
-        return false;
+        return;
     }
 
-    bool added = false;
-    if (!folders.contains(folder_id)) {
-        if (!addCollection(q, collection, folder_id)) {
+    if (!containsFolderId(folders, folder_id)) {
+        if (!addCollection(q, collection, folder_id,
+                           QDateTime() /*last_update*/,
+                           m_embLLM->model() /*embedding_model*/,
```

Comment:
Could we add some code to be certain we don't get here if this comes from addCurrentFolders? In theory that shouldn't happen, but this control flow is heavily dependent on the results of SQL queries that aren't so easy to be _certain_ about. And the last thing we want is to make up an embedding model for an existing collection.

Reply:
fixed.

Reply:
What did you change to fix this?

Reply:
Every place that uses the model name now checks to see if it is empty and if so, it bails out

Reply:
I'm more concerned about the case where the user *does* have a valid embedding model installed and addCurrentFolders calls addFolder - it isn't immediately obvious that this line is absolutely never reachable in that case. containsFolderId should always be true here, so even a `Q_ASSERT` would be good enough, but I really think we should explicitly write that expectation into the code.